grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy

```diff
@@ -0,0 +1,214 @@
+     * @return A collection of the exclude paths that are guaranteed to lead with a '/', e.g [blah, meh].  If null or empty, returns an empty collection
+     * @throws IllegalStateException if any paths are not well-formed to the jobPath
+     */
+    private static Collection<String> createExcludePaths(@Nonnull final String jobPath, @Nullable String paths) {
+        final thisPathsString = paths?.trim()
+        if(!thisPathsString) {
+            return []
+        }
+        final theExcludePaths = thisPathsString.split("\\*") as Collection<String>
+        final jobPathWithSlash = jobPath + '/'
+
+        if (theExcludePaths.any {!it.startsWith(jobPathWithSlash)}) throw new IllegalStateException("Not all exclude paths start with \"${jobPath}\": ${theExcludePaths}")
+
+        return theExcludePaths.collect { it.substring(jobPathWithSlash.length()) }
+    }
+
+    /**
+     * @param jobPath expected to be driven from job parameters
+     * @return a "cleaned" job path, i.e leads with a '/', no trailing '/', and trimmed, e.g "/foo/bar"
+     */
+    private static String cleanJobPath(String jobPath) {
+        String thisJobPath = jobPath.trim()
+        //Add the leading '/' if not present
+        thisJobPath = (thisJobPath[0] != '/') ? "/${thisJobPath}" : thisJobPath
+        //Remove any trailing '/'
+        thisJobPath = (thisJobPath[-1] != '/') ? thisJobPath : thisJobPath[0..-2] as String
+
+        return thisJobPath
+    }
+
+
+    /**
+     * Given the current context in the form of a step contribution, do whatever
+     * is necessary to process this unit inside a transaction. Implementations
+     * return {@link RepeatStatus#FINISHED} if finished. If not they return
+     * {@link RepeatStatus#CONTINUABLE}. On failure throws an exception.
+     *
+     * @param contribution mutable state to be passed back to update the current
+     * step execution
+     * @param chunkContext attributes shared between invocations but not between
+     * restarts
+     * @return an {@link RepeatStatus} indicating whether processing is
+     * continuable.
+     */
+    @Override
+    RepeatStatus execute(StepContribution contribution, ChunkContext chunkContext) throws Exception {
+        final Session session = JcrUtil.getSession(slingRepository, "admin")
+
+        final Node rootNode
+        try {
+            rootNode = session.getNode(jobPath)
+        }
+        catch(PathNotFoundException ex) {
+            log.warn "deleteBeforeWrite was enabled for jobPath ${jobPath} but job path does not exist on client!"
+            return RepeatStatus.FINISHED
+        }
+
+        //If we don't have any exclude paths, we don't have to worry about only clearing particular subtrees
+        if(!relativeExcludePaths || relativeExcludePaths.empty) {
+            rootNode.remove()
+        }
+        else {
+            deletePartialTree(rootNode, relativeExcludePaths)
+        }
+
+        session.save()
```

Comment:
This is just a suggestion (and open for discussion too :) )but you could implement a `DeleteBeforeWriteStepExecutionListener` that prepares (in this case, I guess there's nothing much to prepare; but you can do anything in it) and cleans up after the step (in this case would be `session.save()`)

May be it will be cleaner? :) (example : https://github.com/TWCable/grabbit/blob/master/grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/jcrnodes/JcrNodesStepExecutionListener.groovy )